### PR TITLE
Add background task queue support and use it for CMS image generation

### LINKED ISSUE #14601: Add django-rq support so we can run background tasks
## Description

We'll need a worker queue to pre-process images in the background. Using RQ on Redis to run a worker queue is what we do for Basket and it makes sense to stick with that pattern (plus, of course, it's good enough, too).

---

#### Success Criteria
- [ ] Bedrock supports a redis-backed async queue when Redis is available
- [ ] When redis is NOT available, tasks that normally would go in the background are still executed, just synchronously

### PR DESCRIPTION
**Pull request by stevejalim**
## One-line summary

This changeset adds support for using django-rq as a background task queue (similar to Basket), and then moves our the pre-generation of image "renditions" from CMS-managed images to use that background task queue processing.

- [X] I used an AI to write some of this code: Copliot to write a couple of initial tests, but then had to refactor them  anyway

## Significant changes and points to review

There's a bunch of things going on there, touching our Docker setup, Makefile and our actual app code, so reviewing this commit by commit might be clearest.

Note that while this adds task queue support, it doesn't actually set up the task queue in our infra - that will come as a separate task already listed in #14225

## Issue / Bugzilla link

Resolves #14601

## Testing

* Code review + unit tests passing is probably enough for now. I've already test-driven locally.

### CODE REVIEW

File: bedrock/base/tasks.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    # Avoid mutable default problem
+    func_args = func_args or []
+    func_kwargs = func_kwargs or {}
+
+    if settings.TASK_QUEUE_AVAILABLE:
+        queue = django_rq.get_queue(queue_name)
+        queue.enqueue(func, *func_args, **func_kwargs)
+    else:
+        logger.info("Task queue not available. Immediately executing task")
```
Comment (robhudson) [inline bedrock/base/tasks.py:28]: Might be nice to say which task, or does that already get included in the info?

**Review by robhudson — APPROVED** [review]
I didn't test locally but read closely and couldn't find anything jumping out at me.

File: docker-compose.yml [hunk 2]
```diff
@@ -1,4 +1,3 @@
-version: '3.4'
```
Comment (janbrasna) [inline docker-compose.yml:1]: 👍 Thanks.

This started just recently bugging with [warnings](https://github.com/compose-spec/compose-spec/commit/6d19079ef64dd94d203076e9de6caeb20c498efe), TL;DR: compose 2+ doesn't use it, compose<2 needs it but can't run bedrock due to other things in services definition anyways, so this has no real world impact nowadays.

File: docker/envfiles/local.env [hunk 3]
```diff
@@ -0,0 +1,6 @@
+DEBUG=False
+DEV=False
+ALLOWED_HOSTS=*
+ADMINS=["thedude@example.com"]
+REDIS_URL=redis://redis:6379
+WAGTAIL_ENABLE_ADMIN: True
```
Comment (janbrasna) [inline docker/envfiles/local.env:6]: This will be redefined by the root `.env` that should have the values one would like to use where needed, so just a naïve Q: do the `DEV`+`DEBUG` have a reason to be `False` instead of `True` here? Is it so that the worker job itself shouldn't need to print out any debug info per se?

File: docs/task-queue.rst [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+.. License, v. 2.0. If a copy of the MPL was not distributed with this
+.. file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+.. _task-queue:
+
+==========
+Task queue
+==========
+
+As part of the 2024 Wagtail CMS work, have added a task queue to process
```
Comment (janbrasna) [inline docs/task-queue.rst:11]:
```suggestion
As part of the 2024 Wagtail CMS work, we have added a task queue to process
```
?

File: docs/task-queue.rst [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+longer-running jobs in the background.
+
+The queue is supplied by ``django-rq``, which sits on top of ``rq`` and uses
+Redis as the storage backend.
+
+Once infra has been set up fully to support this, documentation will be expanded
```
Comment (janbrasna) [inline docs/task-queue.rst:17]:
```suggestion
Once infra has been set up fully to support this, documentation will be expanded,
```

**Review by janbrasna — COMMENTED** [review]
Only formatting nits from quickly glancing the docs:

File: docker/envfiles/local.env (hunk 3 above)
Comment (stevejalim) [reply docker/envfiles/local.env:6]: Yeah, I guess it could work in Debug mode locally. My default is to keep things strict in order to avoid surprises, unless there's a need to relax it, but happy to tweak.

Comment (stevejalim) [reply docker/envfiles/local.env:6]: Actually, thinking about this a bit more, keeping it stricter by default will mean running `release-local` and the local queue behave as in production. AFAICT `DEBUG` doesn't directly influence our logging levels, but does make Django's internals behave differently upon error, and I'd rather - by default - see prod-like problems than debug-mode versions of them. As you say, this will be overridden by anything in an on-filesystem `.env` file

**Comment by janbrasna** [comment]
@stevejalim I'm probably missing something, but the worker service keeps failing/restarting for me, basically taking all the computing power just starting over and over. No amount of make clean or make pulls make any difference. (It does `docker-compose pull release app assets builder app-base` that's sufficient right?)

I see it prints there are 3 migrations to run, do I need to do that inside the docker shell of one of the services for it to work? Or it is unrelated?

I get gazillions of:
Screenshot 2024-06-17 at 16 02 13
Screenshot 2024-06-17 at 16 02 31

Basically it ends up in a state when containers can't be stopped or killed, only shutting down the engine completely helps to cool it down;) Disclaimer: I don't have access to uptodate docker version, I have to stick with desktop 4.15.0 for several more weeks so sorry about not being able to test repro on another darwin+engine combo:/

**Comment by stevejalim** [comment]
Ah, i think the issue here is that the DB lacks the product_details table and you end up in a crashloop as a result

**Comment by stevejalim** [comment]
@janbrasna as a quick fix for now I'll remove the worker from the docker compose config

**Comment by janbrasna** [comment]
@stevejalim don't, if you don't see a reason for this — I happily use `docker-compose up app assets` instead for now without any obvious issues.

**Comment by stevejalim** [comment]
@janbrasna well others may use the makefile too and it looks like a foot gun that I need to fix

**Comment by stevejalim** [comment]
https://github.com/mozilla/bedrock/pull/14694
